bundles/core/src/main/java/com/adobe/cq/commerce/core/search/internal/services/SearchResultsServiceImpl.java

```diff
@@ -141,20 +143,34 @@ public SearchResultsSet performSearch(
             return new ImmutablePair<>(null, searchResultsSet);
         }
 
+        // Next we generate the graphql category query and actually query the commerce system
+        CategoryTree category = null;
+        if (generateCategoryQueryString(categoryRetriever).isPresent()) {
+            String categoryQueryString = generateCategoryQueryString(categoryRetriever).get();
+            LOGGER.debug("Generated category query string {}", categoryQueryString);
+            GraphqlResponse<Query, Error> categoryResponse = magentoGraphqlClient.execute(categoryQueryString);
+            Query categoryData = categoryResponse.getData();
+
+            if (categoryData != null) {
+                List<CategoryTree> categories = categoryData.getCategoryList();
+                if (CollectionUtils.isNotEmpty(categories)) {
+                    category = categories.get(0);
+                    ((SearchOptionsImpl) searchOptions).setCategoryId(category.getId().toString());
```

Comment:
The SearchResultsService is public API, hard casting here implicitly introduces a dependency to an un-exported package. (the one containing the SearchOptionsImpl).

I would recommend to add a constructor `SearchOptionsImpl(SearchOptions)`. It creates a 1:1 copy of the given SearchOptions as `SearchOptionsImpl`, which you can then mutate. 

Not to handle in this PR, but SearchOptions should be DTO w/o an interface and the logic should be kept in a service implementation.